nir/src/main/scala/scala/scalanative/nir/Fresh.scala

```diff
@@ -3,13 +3,31 @@ package nir
 
 import java.util.concurrent.atomic.AtomicInteger
 
-final class Fresh(scope: String) {
-  private var i = new AtomicInteger(0)
-  def apply() = {
-    val value = i.getAndIncrement()
-    Local(scope, value)
+final class Fresh private (private var start: Int) {
+  def apply(): Local = {
+    start += 1
+    val value = start
+    Local(value)
   }
 }
+
 object Fresh {
-  def apply(scope: String) = new Fresh(scope)
+  def apply(start: Int = 0): Fresh =
+    new Fresh(start)
+
+  def apply(insts: Seq[Inst]): Fresh = {
+    var max = -1
+    insts.foreach {
+      case Inst.Let(local, _) =>
+        max = Math.max(max, local.id)
+      case Inst.Label(local, params) =>
+        max = Math.max(max, local.id)
+        params.foreach { param =>
+          max = Math.max(max, param.name.id)
+        }
+      case _ =>
+        ()
+    }
+    new Fresh(max)
+  }
```

Comment:
It seems like performance overhead is negligible: 

optimizeNIR on unit-tests, before:

```

[info] Optimizing (2814 ms)
[info] Optimizing (2056 ms)
[info] Optimizing (1648 ms)
[info] Optimizing (1729 ms)
[info] Optimizing (2055 ms)
```

after:
```
[info] Optimizing (2801 ms)
[info] Optimizing (2059 ms)
[info] Optimizing (1763 ms)
[info] Optimizing (1686 ms)
[info] Optimizing (1682 ms)
```